hardware_interface/include/hardware_interface/joint.hpp

```diff
@@ -0,0 +1,361 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef HARDWARE_INTERFACE__JOINT_HPP_
+#define HARDWARE_INTERFACE__JOINT_HPP_
+
+#include <string>
+#include <vector>
+
+#include "hardware_interface/component_info.hpp"
+#include "hardware_interface/hardware_info.hpp"
+#include "hardware_interface/types/hardware_interface_return_values.hpp"
+#include "hardware_interface/visibility_control.h"
+
+namespace hardware_interface
+{
+
+/**
+ * \brief Base Class for the "Joint" component used as a basic building block for a robot.
+ * A joint is always 1-DoF and can have one or more interfaces (e.g., position, velocity, etc.)
+ * A joint has to be able to receive command(s) and optionally can provide its state(s).
+ */
+class Joint
+{
+public:
+  HARDWARE_INTERFACE_PUBLIC
+  Joint() = default;
+
+  HARDWARE_INTERFACE_PUBLIC
+  virtual
+  ~Joint() = default;
+
+  /**
+   * \brief Configure joint based on the description in the robot's URDF file.
+   *
+   * \param joint_info structure with data from URDF.
+   * \return return_type::OK if required data are provided and is successfully parsed,
+   * return_type::ERROR otherwise.
+   */
+  HARDWARE_INTERFACE_PUBLIC
+  virtual
+  return_type configure(const ComponentInfo & joint_info) = 0;
+
+  /**
+   * \brief Provide the list of command interfaces configured for the joint.
+   *
+   * \return string list with command interfaces.
+   */
+  HARDWARE_INTERFACE_PUBLIC
+  std::vector<std::string> get_command_interfaces() const
+  {
+    return info_.command_interfaces;
+  }
+
+  /**
+   * \brief Provide the list of state interfaces configured for the joint.
+   *
+   * \return string list with state interfaces.
+   */
+  HARDWARE_INTERFACE_PUBLIC
+  std::vector<std::string> get_state_interfaces() const
+  {
+    return info_.state_interfaces;
+  }
+
+  /**
+   * \brief Get command list from the joint. This function is used in the write function of the
+   * actuator or system hardware. The parameters command and interfaces have the same order,
+   * and number of elements. Using the interfaces list, the hardware can choose which values to
+   * provide.
+   *
+   * \param command list of doubles with commands for the hardware.
+   * \param interfaces list of interfaces on which commands have to set.
+   * \return return_type::INTERFACE_VALUE_SIZE_NOT_EQUAL if command and interfaces arguments do not
+   * have the same length; return_type::INTERFACE_NOT_FOUND if one of provided interfaces is not
+   * defined for the joint; return_type::OK otherwise.
```

Comment:
```suggestion
   * defined for the joint; return return_type::INTERFACE_NOT_PROVIDED if the list of interfaces is empty; return_type::OK otherwise.
```